@tailwindcss/vite: build error when use @plugin "@tailwindcss/typography";

**What version of Tailwind CSS are you using?**

v4.2.3

**What build tool (or framework if it abstracts the build tool) are you using?**

Vite 8.0.9

**What version of Node.js are you using?**

v24.15.0

**What operating system are you using?**

macOS


when used together with resolve.alias in vite.config

```
resolve: {
    alias: [
      { find: '@', replacement: fileURLToPath(new URL('.', import.meta.url)) },
    ],
  },
```

I can reproduce this issue. Using Linux (Fedora 43 Workstation), SvelteKit 2.57.1, Node 25.9.0, Vite 8.0.9, and Tailwind 4.2.3, the project fails to build:
```
✓ 226 modules transformed.
✗ Build failed in 38.07s
error during build:
Build failed with 1 error:

[plugin @tailwindcss/vite:generate:build] <project directory>/src/routes/layout.css
Error: Cannot find module '<project directory>/@tailwindcss/typography'
```

Reverting Tailwind (and @tailwindcss/vite) to 4.2.2 resolves this issue.

Same issue here after upgrading to `4.2.3`. Running TanStack Start with Vite `8.0.9` and rolling back to `4.2.2` sorts it. This is the error I get when running `npm run build` locally:
```bash
✓ 225 modules transformed.
✗ Build failed in 390ms
error during build:
Build failed with 1 error:

[plugin @tailwindcss/vite:generate:build] C:/Users/<user>/Repositrories/<repo>/src/styles/global.css?transform-only
Error: Cannot find module 'C:\Users\<user>\Repositrories\<repo>\@tailwindcss\typography'   
Require stack:
- C:\Users\<user>\Repositrories\<repo>\node_modules\@tailwindcss\node\dist\index.mjs       
    at Module._resolveFilename (node:internal/modules/cjs/loader:1448:15)
    at require.resolve (node:internal/modules/helpers:163:19)
    at jitiResolve (C:\Users\<user>\Repositrories\<repo>\node_modules\jiti\dist\jiti.cjs:1:148703)
    at jitiRequire (C:\Users\<user>\Repositrories\<repo>\node_modules\jiti\dist\jiti.cjs:1:150290)
    at Function.import (C:\Users\<user>\Repositrories\<repo>\node_modules\jiti\dist\jiti.cjs:1:158307)
    at vt (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/@tailwindcss/node/dist/index.mjs:10:4506)
    at async xt (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/@tailwindcss/node/dist/index.mjs:10:3774)
    at async file:///C:/Users/<user>/Repositrories/<repo>/node_modules/tailwindcss/dist/chunk-QYH7VNSF.mjs:27:1070
    at async Promise.all (index 0)
    at async Promise.all (index 1)
    at aggregateBindingErrorsIntoJsError (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/rolldown/dist/shared/error-DAA7ncC5.mjs:48:18)
    at unwrapBindingResult (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/rolldown/dist/shared/error-DAA7ncC5.mjs:18:128)
    at #build (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/rolldown/dist/shared/rolldown-build-BPKCFYpX.mjs:3317:34)
    at async buildEnvironment (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/vite/dist/node/chunks/node.js:32999:64)
    at async Object.build (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/vite/dist/node/chunks/node.js:33421:19)
    at async buildStartViteEnvironments (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/@tanstack/start-plugin-core/dist/esm/vite/planning.js:94:23)
    at async Object.buildApp (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/@tanstack/start-plugin-core/dist/esm/vite/plugin.js:110:8)
    at async Object.buildApp (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/vite/dist/node/chunks/node.js:33413:6)
    at async CAC.<anonymous> (file:///C:/Users/<user>/Repositrories/<repo>/node_modules/vite/dist/node/cli.js:778:3) {
  errors: [Getter/Setter]
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - _Experimental_: add `@container-size` utility ([#18901](https://github.com/tailwindlabs/tailwindcss/pull/18901))
 
+### Fixed
+
+- Ensure imports in `@import` and `@plugin` still resolve correctly when using Vite aliases in `@tailwindcss/vite` ([#19947](https://github.com/tailwindlabs/tailwindcss/pull/19947))
+
 ## [4.2.3] - 2026-04-20
 
 ### Fixed

--- a/integrations/vite/resolvers.test.ts
+++ b/integrations/vite/resolvers.test.ts
@@ -164,6 +164,130 @@ test(
   },
 )
 
+test(
+  'resolves at-sign aliases in production build',
+  {
+    fs: {
+      'package.json': json`
+        {
+          "type": "module",
+          "dependencies": {
+            "@tailwindcss/vite": "workspace:^",
+            "tailwindcss": "workspace:^"
+          },
+          "devDependencies": {
+            "vite": "^8"
+          }
+        }
+      `,
+      'vite.config.ts': ts`
+        import tailwindcss from '@tailwindcss/vite'
+        import { defineConfig } from 'vite'
+        import { fileURLToPath } from 'node:url'
+
+        export default defineConfig({
+          build: { cssMinify: false },
+          plugins: [tailwindcss()],
+          resolve: {
+            alias: [{ find: '@', replacement: fileURLToPath(new URL('.', import.meta.url)) }],
+          },
+        })
+      `,
+      'index.html': html`
+        <head>
+          <link rel="stylesheet" href="./src/index.css" />
+        </head>
+        <body>
+          <div class="underline custom-underline">Hello, world!</div>
+        </body>
+      `,
+      'src/index.css': css`
+        @import '@/src/styles/base.css';
+        @plugin '@/src/plugin.js';
+      `,
+      'src/styles/base.css': css`
+        @reference 'tailwindcss/theme';
+        @import 'tailwindcss/utilities';
+      `,
+      'src/plugin.js': js`
+        export default function ({ addUtilities }) {
+          addUtilities({ '.custom-underline': { 'border-bottom': '1px solid green' } })
+        }
+      `,
+    },
+  },
+  async ({ fs, exec, expect }) => {
+    await exec('pnpm vite build')
+
+    let files = await fs.glob('dist/**/*.css')
+    expect(files).toHaveLength(1)
+    let [filename] = files[0]
+
+    await fs.expectFileToContain(filename, [candidate`underline`, candidate`custom-underline`])
+  },
+)
+
+test(
+  'resolves package plugins in production build with at-sign aliases',
+  {
+    fs: {
+      'package.json': json`
+        {
+          "type": "module",
+          "dependencies": {
+            "@tailwindcss/vite": "workspace:^",
+            "tailwindcss": "workspace:^"
+          },
+          "devDependencies": {
+            "tailwind-scrollbar": "^4.0.2",
+            "tailwindcss-motion": "^1.1.1",
+            "vite": "^8"
+          }
+        }
+      `,
+      'vite.config.ts': ts`
+        import tailwindcss from '@tailwindcss/vite'
+        import { defineConfig } from 'vite'
+        import { fileURLToPath } from 'node:url'
+
+        export default defineConfig({
+          build: { cssMinify: false },
+          plugins: [tailwindcss()],
+          resolve: {
+            alias: [{ find: '@', replacement: fileURLToPath(new URL('.', import.meta.url)) }],
+          },
+        })
+      `,
+      'index.html': html`
+        <head>
+          <link rel="stylesheet" href="./src/index.css" />
+        </head>
+        <body>
+          <div class="scrollbar scrollbar-thumb-red-500 motion-preset-fade">Hello, world!</div>
+        </body>
+      `,
+      'src/index.css': css`
+        @import 'tailwindcss';
+        @plugin 'tailwind-scrollbar';
+        @plugin 'tailwindcss-motion';
+      `,
+    },
+  },
+  async ({ fs, exec, expect }) => {
+    await exec('pnpm vite build')
+
+    let files = await fs.glob('dist/**/*.css')
+    expect(files).toHaveLength(1)
+    let [filename] = files[0]
+
+    await fs.expectFileToContain(filename, [
+      candidate`scrollbar`,
+      candidate`scrollbar-thumb-red-500`,
+      candidate`motion-preset-fade`,
+    ])
+  },
+)
+
 describe.each(['postcss', 'lightningcss'])('%s', (transformer) => {
   test(
     'resolves aliases in production build',

--- a/integrations/vite/sveltekit.test.ts
+++ b/integrations/vite/sveltekit.test.ts
@@ -1,0 +1,80 @@
+import { candidate, css, html, js, json, test, ts } from '../utils'
+
+test(
+  'sveltekit build resolves package plugins',
+  {
+    fs: {
+      'package.json': json`
+        {
+          "type": "module",
+          "scripts": {
+            "prepare": "svelte-kit sync || echo ''"
+          },
+          "dependencies": {
+            "svelte": "^5",
+            "tailwindcss": "workspace:^"
+          },
+          "devDependencies": {
+            "@sveltejs/adapter-auto": "^6",
+            "@sveltejs/kit": "^2",
+            "@tailwindcss/forms": "^0.5.11",
+            "@tailwindcss/vite": "workspace:^",
+            "vite": "^8"
+          }
+        }
+      `,
+      'svelte.config.js': js`
+        import adapter from '@sveltejs/adapter-auto'
+
+        export default {
+          kit: {
+            adapter: adapter(),
+          },
+        }
+      `,
+      'vite.config.ts': ts`
+        import tailwindcss from '@tailwindcss/vite'
+        import { sveltekit } from '@sveltejs/kit/vite'
+        import { defineConfig } from 'vite'
+
+        export default defineConfig({
+          build: { cssMinify: false },
+          plugins: [tailwindcss(), sveltekit()],
+        })
+      `,
+      'src/app.html': html`
+        <!doctype html>
+        <html lang="en">
+          <head>
+            %sveltekit.head%
+          </head>
+          <body data-sveltekit-preload-data="hover">
+            <div style="display: contents">%sveltekit.body%</div>
+          </body>
+        </html>
+      `,
+      'src/routes/+layout.svelte': html`
+        <script>
+          import '../app.css'
+        </script>
+
+        <slot />
+      `,
+      'src/routes/+page.svelte': html`
+        <div class="form-input">Hello, world!</div>
+      `,
+      'src/app.css': css`
+        @import 'tailwindcss';
+        @plugin '@tailwindcss/forms';
+      `,
+    },
+  },
+  async ({ exec, fs, expect }) => {
+    await exec('pnpm vite build')
+
+    let files = await fs.glob('.svelte-kit/output/client/**/*.css')
+    expect(files.length).toBeGreaterThan(0)
+
+    await fs.expectFileToContain(files[0][0], [candidate`form-input`])
+  },
+)

--- a/packages/@tailwindcss-postcss/README.md
+++ b/packages/@tailwindcss-postcss/README.md
@@ -109,4 +109,3 @@ export default {
   ],
 }
 ```
-

--- a/packages/@tailwindcss-vite/src/index.ts
+++ b/packages/@tailwindcss-vite/src/index.ts
@@ -64,10 +64,19 @@ export default function tailwindcss(opts: PluginOptions = {}): Plugin[] {
 
       customCssResolver = async (id: string, base: string) => {
         let resolved = await cssResolver(id, base, false, isSSR)
-        if (resolved && !resolved.endsWith('.css')) return undefined
+        if (!resolved) return
+        if (resolved === id) return
+        if (!path.isAbsolute(resolved)) return
+        if (!resolved.endsWith('.css')) return
         return resolved
       }
-      customJsResolver = (id: string, base: string) => jsResolver(id, base, false, isSSR)
+      customJsResolver = async (id: string, base: string) => {
+        let resolved = await jsResolver(id, base, false, isSSR)
+        if (!resolved) return
+        if (resolved === id) return
+        if (!path.isAbsolute(resolved)) return
+        return resolved
+      }
     } else {
       type ResolveIdFn = (
         environment: Environment,
@@ -111,10 +120,19 @@ export default function tailwindcss(opts: PluginOptions = {}): Plugin[] {
 
       customCssResolver = async (id: string, base: string) => {
         let resolved = await cssResolver(env, id, base, false)
-        if (resolved && !resolved.endsWith('.css')) return undefined
+        if (!resolved) return
+        if (resolved === id) return
+        if (!path.isAbsolute(resolved)) return
+        if (!resolved.endsWith('.css')) return
         return resolved
       }
-      customJsResolver = (id: string, base: string) => jsResolver(env, id, base, false)
+      customJsResolver = async (id: string, base: string) => {
+        let resolved = await jsResolver(env, id, base, false)
+        if (!resolved) return
+        if (resolved === id) return
+        if (!path.isAbsolute(resolved)) return
+        return resolved
+      }
     }
 
     return new Root(

